src/coreclr/nativeaot/Runtime/eventpipe/ep-rt-aot.cpp

```diff
@@ -85,12 +84,70 @@ ep_rt_aot_sample_profiler_write_sampling_event_for_threads (
 {
 }
 
+#if defined (__linux__)
+extern char *__progname;
+#endif
+
 const ep_char8_t *
 ep_rt_aot_entrypoint_assembly_name_get_utf8 (void) 
 {
-    // shipping criteria: no EVENTPIPE-NATIVEAOT-TODO left in the codebase
-    // TODO: Implement EventPipe assembly name - return filename in nativeaot?
-    return reinterpret_cast<const ep_char8_t*>("");
+    // Although PalGetModuleFileName exists, non-windows implementation has issues and we use __progname for Linux.
+    // Cannot use __cpp_threadsafe_static_init feature since it will bring in the C++ runtime and need to use threadsafe way to initialize entrypoint_assembly_name
+    static const ep_char8_t * entrypoint_assembly_name = nullptr;
+    if (entrypoint_assembly_name == nullptr) {
+        ep_char8_t * entrypoint_assembly_name_local;
+#ifdef HOST_WINDOWS
+        const TCHAR * wszModuleFileName = NULL;
+        if(PalGetModuleFileName(&wszModuleFileName, nullptr) == 0) {
+            // Allocating since we delete this empty string later
+            entrypoint_assembly_name_local = reinterpret_cast<ep_char8_t *>(malloc(1));
+            entrypoint_assembly_name_local = reinterpret_cast<ep_char8_t *>('\0');
+        } else {
+            const wchar_t* process_name = wcsrchr(wszModuleFileName, '\\');
+            if (process_name != NULL) {
+                // We don't want to include the first '\'
+                process_name++;
+                const wchar_t* extension = wcschr(process_name, '.');
+                if (extension != NULL) {
+                    size_t len = extension - process_name;
+                    const ep_char16_t* process_name_wo_ext_l = reinterpret_cast<const ep_char16_t *>(process_name);
+                    entrypoint_assembly_name_local = ep_rt_utf16_to_utf8_string(process_name_wo_ext_l, len);
+                } else {
+                    const ep_char16_t* process_name_l = reinterpret_cast<const ep_char16_t *>(process_name);
+                    entrypoint_assembly_name_local = ep_rt_utf16_to_utf8_string(process_name_l, -1);
+                }
+            } else{
+                // Allocating since we delete this empty string later
+                entrypoint_assembly_name_local = reinterpret_cast<ep_char8_t *>(malloc(1));
+                entrypoint_assembly_name_local = reinterpret_cast<ep_char8_t *>('\0');
```

Comment:
This line throws away the pointer that was just allocated. Should this initialize the memory that was just allocated instead?

Also, this should handle allocation failure.